--- kadm5-0.2.3.org/kadm5.c	2003-05-28 20:51:42.000000000 +1000
+++ kadm5-0.2.3/kadm5.c	2012-02-20 20:52:25.958767002 +1100
@@ -104,7 +104,7 @@
 	NULL,
 	PHP_MINFO(kadm5),
 #if ZEND_MODULE_API_NO >= 20010901
-	"0.2.3",
+	"0.2.4",
 #endif
 	STANDARD_MODULE_PROPERTIES
 };
@@ -436,6 +436,7 @@
 		php_error(E_WARNING, "Multiple values for single or folded enctype. (KADM5_SETKEY_DUP_ENCTYPES)");
 		break;
 	default:
+		break;
 	}
 }
 /* }}} */
@@ -450,6 +451,9 @@
 	kadm5_config_params params;
 	void *handle = NULL;
 	kadm5_ret_t rc;
+	krb5_error_code rv;
+	krb5_context ctx;
+	char **db_args = NULL;
 
 	memset((char *) &params, 0, sizeof(params));
 
@@ -465,13 +469,24 @@
 	params.admin_server = admin_server;
 	params.mask |= KADM5_CONFIG_ADMIN_SERVER;
 
-	rc = kadm5_init_with_password(princstr,
-								  password,
-								  KADM5_ADMIN_SERVICE,
-								  &params,
-								  KADM5_STRUCT_VERSION,
-								  KADM5_API_VERSION_2,
-								  &handle);
+	rv = krb5_init_context(&ctx);
+	if (rv) {
+	// should handle using set_com_err_hook for full errors
+        //com_err("PHP KADM5", retval, "while initializing krb5"); 
+		php_error(E_WARNING, "Error while initializing krb5 library");
+		RETURN_FALSE;        
+	}
+
+	rc = kadm5_init_with_password(
+	ctx,
+	princstr,
+	password,
+	KADM5_CHANGEPW_SERVICE,
+	&params, // struct of server, host, etc 
+	KADM5_STRUCT_VERSION,
+	KADM5_API_VERSION_3,
+	db_args,
+	&handle);
 
 	if (rc) {
 		kadm5_error(rc);
@@ -494,14 +509,20 @@
 {
 	zval **link;
 	void *handle;
+	kadm5_ret_t rc;
 
-	if (ZEND_NUM_ARGS() != 1 || zend_get_parameters_ex(1, &link ) == FAILURE) {
-		WRONG_PARAM_COUNT;
-	}
+	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "z", &link)
+	 == FAILURE)
+		RETURN_NULL();
+
+	ZEND_FETCH_RESOURCE(handle, void *, &link, -1, HANDLE_ID, le_handle);
+	rc = kadm5_destroy(handle);
 
-	ZEND_FETCH_RESOURCE(handle, void *, link, -1, HANDLE_ID, le_handle);
+	if (rc) {
+		kadm5_error(rc);
+		RETURN_FALSE;
+	}
 
-	zend_list_delete((*link)->value.lval);
 	RETURN_TRUE;
 }
 /* }}} */
@@ -515,12 +536,11 @@
 	void *handle;
 	kadm5_ret_t rc;
 
-	if (ZEND_NUM_ARGS() != 1 || zend_get_parameters_ex(1, &link ) == FAILURE) {
-		WRONG_PARAM_COUNT;
-	}
-
-	ZEND_FETCH_RESOURCE(handle, void *, link, -1, HANDLE_ID, le_handle);
+	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "z", &link)
+	 == FAILURE)
+		RETURN_NULL();
 
+	ZEND_FETCH_RESOURCE(handle, void *, &link, -1, HANDLE_ID, le_handle);
 	rc = kadm5_flush(handle);
 
 	if (rc) {
