--- kadm5-0.2.3.org/kadm5.c	2003-05-28 20:51:42.000000000 +1000
+++ kadm5-0.2.3/kadm5.c	2012-02-14 12:49:41.170545001 +1100
@@ -436,6 +436,7 @@
 		php_error(E_WARNING, "Multiple values for single or folded enctype. (KADM5_SETKEY_DUP_ENCTYPES)");
 		break;
 	default:
+		break;
 	}
 }
 /* }}} */
@@ -450,6 +451,9 @@
 	kadm5_config_params params;
 	void *handle = NULL;
 	kadm5_ret_t rc;
+	krb5_error_code rv;
+	krb5_context ctx;
+	char **db_args = NULL;
 
 	memset((char *) &params, 0, sizeof(params));
 
@@ -465,16 +469,28 @@
 	params.admin_server = admin_server;
 	params.mask |= KADM5_CONFIG_ADMIN_SERVER;
 
-	rc = kadm5_init_with_password(princstr,
-								  password,
-								  KADM5_ADMIN_SERVICE,
-								  &params,
-								  KADM5_STRUCT_VERSION,
-								  KADM5_API_VERSION_2,
-								  &handle);
+	rv = krb5_init_context(&ctx);
+	if (rv) {
+	// should handle using set_com_err_hook for full errors
+        //com_err("PHP KADM5", retval, "while initializing krb5"); 
+		php_error(E_WARNING, "Error while initializing krb5 library");
+		RETURN_FALSE;        
+	}
+
+	rc = kadm5_init_with_password(
+	ctx,
+	princstr,
+	password,
+	KADM5_ADMIN_SERVICE,
+	&params, // struct of server, host, etc 
+	KADM5_STRUCT_VERSION,
+	KADM5_API_VERSION_3,
+	db_args,
+	&handle);
 
 	if (rc) {
 		kadm5_error(rc);
+		//com_err("PHP KADM5", rc, "while doing password init.");
 		RETURN_FALSE;
 	}
 
